Mech camera

Sealed101 opened the pull request:
## About The Pull Request
This PR implements a new add-on for mechs that functions as a ~~EMP-proof~~ borg security camera. Unlocked in Basic Robotics Research node.
Mechs with a camera installed will have a unique diagnostic HUD blip.

![image](https://user-images.githubusercontent.com/75863639/202909177-f2badafc-2429-495f-a281-fcaad95837fb.png)

Getting EMPd temporarily sets the camera's view range to 1.

![image](https://user-images.githubusercontent.com/75863639/203862438-cfcdc442-3aa6-4468-83bb-b4ccad8dfa2e.png)


## Why It's Good For The Game
Bodycam footage of the mech in action; AIs piloting a mech with this add-on can actually see what's happening around the mech, much like following a borg.

## Changelog
:cl:
add: robotics now has access to a mech camera add-on, which installs a security camera into the mech. bodycam footage from the big leagues!
qol: cyborg cameras should be more active with updating static when moving (ideally down from 2.5 seconds to 0.5 between updates), make sure to report any jank if you happen on some
/:cl:

Sealed101 committed:
initial
atm camera autonaming is the only thing that works

Sealed101 committed:
i need to go to bed NOW

Sealed101 committed:
hud fix, camera preset & camera kit repath
bonus: codersprites

Sealed101 committed:
Merge branch 'tgstation:master' into mech-camera

Sealed101 committed:
hjonk

Sealed101 committed:
Merge branch 'tgstation:master' into mech-camera

Sealed101 committed:
added to exosuit fab
update cd to 0.5 sec

Sealed101 committed:
Merge branch 'master' into mech-camera

Sealed101 committed:
some autoname stuff

Sealed101 committed:
bit more autoname stuff

Sealed101 committed:
whoops

Sealed101 commented:
fun fact: static updates are tied to this cooldown `UPDATE_BUFFER_TIME` of 2.5 seconds
https://github.com/tgstation/tgstation/blob/eef9ab04bcb945bced196b34322a3c8cedf71755/code/modules/mob/living/silicon/ai/freelook/chunk.dm#L60-L64

which makes the following silicon and mech procs to have a seemingly pointless cooldown
https://github.com/tgstation/tgstation/blob/eef9ab04bcb945bced196b34322a3c8cedf71755/code/modules/vehicles/mecha/mecha_movement.dm#L168-L174

https://github.com/tgstation/tgstation/blob/eef9ab04bcb945bced196b34322a3c8cedf71755/code/modules/mob/living/silicon/silicon_movement.dm#L16-L22

Dorsisdwarf commented:
Sec officer bodycams were great fun to watch (if unbalanced) so I expect this to also be cool.

optimumtact commented:
> EMP-proof camera. 

why EMP-proof

OceanFish1 commented:
I think EMP-proofness must go

Sealed101 commented:
emp proof because any source of EMP on the mech like installing both this camera and a mech tracker would lead to the camera being disabled by the tracker's EMP, which will cut off any surveillance over the mech. that can be a problem for the party that's trying to stop the mech, if it doesn't have a constant line of sight of it. of course, it is also be a problem for the piloting AIs, but cutting off their camera feed in a place with no camera coverage can basically mean game over. camera would get disabled for a long period of time, leaving the AI pilot stumbling in the camera static. that just seems like an unfair scenario that a silicon pilot would not have any counterplay against, while a carbon pilot would still be able to walk away

optimumtact commented:
> that can be a problem for the party that's trying to stop the mech, if it doesn't have a constant line of sight of it. 

The same party that can remotely EMP it? Doesn't really make sense

>it is also be a problem for the piloting AIs, but cutting off their camera feed in a place with no camera coverage can basically mean game over. camera would get disabled for a long period of time, leaving the AI pilot stumbling in the camera static.

if I understand this correctly currently AI's can only pilot a mech where they have an active camera network, is that correct?

While i'm open to the arguments about the AI, it seems like it would be better suited to make it have a timed outage period (not permanent), but also not completely immune either.

Sealed101 commented:
> The same party that can remotely EMP it? Doesn't really make sense

Yes and by triggering the beacon's EMP they would rob themselves and any other camera users of the camera feed, which would sacrifice remote surveillance of the mech. This could lead to both parties being left without any camera coverage in areas that lack a camera network, like station space and maints.

> if I understand this correctly currently AI's can only pilot a mech where they have an active camera network, is that correct?
> 
> While i'm open to the arguments about the AI, it seems like it would be better suited to make it have a timed outage period (not permanent), but also not completely immune either.

An AI can pilot a mech in static, but without a camera network it has no way to see what's happening, i. e. space/lavaland environments. 

____

That said, I've discovered that mechs handling EMP seems to currently be broken, as they EMP-proof their contents completely with no system like empprotection element behind it, and that is including the pilot. I was under the impression that this wasn't the case, that's my bad. I am looking into and will open a separate PR for the issue.

I will also see how would a momentary cutoff of the camera feed work, but for now, this is a draft.

Sealed101 committed:
Merge branch 'master' into mech-camera

Sealed101 committed:
what if it messed with the focus instead
un-EMP-proofs the camera
EMPs mess up the camera's view range to 1
lasts for (10/severity) seconds, duration not refreshed if EMPd again
also minor comment fix in cameranet (im stupipd)

Sealed101 committed:
scrubs mentions of EMP-proof & adds EMPd HUD image

Sealed101 committed:
cleanup

Sealed101 commented:
Now when EMPd, camera gets it's view range set to 1 tile. It will reset after 10 seconds / severity of the emp (so a light emp that's flagged as 2 will reset the camera after 5 seconds). I think this works better than a complete blackout, since it gives the AI a little bit of leeway for movement or melee-range interactions while not shutting off camera feed for anyone watching.

The current issue is that active camera console sessions locked onto the mech camera will not update the range of the view feed (at least while the mech isn't moving). It also works fucking horrible with the delay mentioned in the first comment on the PR

<details>
<summary>With 2.5 seconds camera update delay<summary/>


https://user-images.githubusercontent.com/75863639/203947343-f8ddfc2d-b2f4-499b-a456-c0533bc227f1.mp4


</details>


______

<details>
<summary>With mech camera's 0.5 seconds camera update delay<summary/>


https://user-images.githubusercontent.com/75863639/203947492-711e2e02-86cb-4747-8d6a-ddb5cc5e4d2a.mp4


</details>

Time-Green commented on code/game/machinery/camera/presets.dm:
```diff
@@ -66,6 +66,37 @@
+	desc = "This camera belongs in a mecha. If you see this, tell a coder!"
+	network = list("ss13", "rd")
+	short_range = 1 //used when the camera gets EMPd
+	var/number
```
Whats number? Autodoc this

Time-Green commented on code/game/machinery/computer/camera.dm:
```diff
@@ -132,7 +132,7 @@
 
 	// Is this camera located in or attached to a living thing? If so, assume the camera's loc is the living thing.
-	var/cam_location = isliving(active_camera.loc) ? active_camera.loc : active_camera
+	var/cam_location = (isliving(active_camera.loc) || ismecha(active_camera.loc)) ? active_camera.loc : active_camera
```
What's the point of these if-checks? We only ever use cam_location inside a get_turf, and get_turf always gets the turf regardless of how many containers its in

Sealed101 replied:
iirc this makes it so the camera feed isn't taken from the camera itself(objects inside another object have z = 0, so that makes a blackscreen instead of the camera feed) but the object it's in

Time-Green replied:
Yeah it grabs the turf, so why check for isliving and ismecha? If you scroll down a bit you can see it always grabs the turf, so it's pointless

Time-Green commented on code/modules/vehicles/mecha/mecha_defense.dm:
```diff
@@ -168,6 +168,13 @@
+		chassis_camera.setViewRange(chassis_camera.short_range)
+		chassis_camera.is_emp_scrambled = TRUE
+		diag_hud_set_camera()
+		addtimer(CALLBACK(chassis_camera, TYPE_PROC_REF(/obj/machinery/camera/exosuit, emp_refocus), src), 10 SECONDS / severity)
```
Don't make cameras EMP-proof if they're already EMP'd, it's a bit unintuitive

```suggestion
		addtimer(CALLBACK(chassis_camera, TYPE_PROC_REF(/obj/machinery/camera/exosuit, emp_refocus), src), 10 SECONDS / severity, TIMER_OVERRIDE | TIMER_UNIQUE)
```
The above makes it so the timer overrides the previous timer if you call addtimer again while your timer is already ticking, so you don't get emp_refocus called out of sequence.

Sealed101 replied:
i was actually kind of going for the emp effect's duration to not be refreshed on successful emps, but sure

Time-Green replied:
Oh, I mean sure. Can you explain why you think its a good idea? I dont see it

Sealed101 replied:
continuous EMPs only really come from ion bolts and those are pretty much always just emptied into the mech. this would make the camera stuck to the debuffed view_range for the entire duration of the firefight, which while inconsequential for a carbon mech pilot, it would cripple an AI pilot's FoV completely. having it not refreshed on consecutive EMPs can provide at least some info on the attackers for an AI pilot before getting EMPd and shortsighted again

tl; dr - chaining EMPs to almost completely blind an AI piloted mech is a huge disadvantage. if it's not refreshed on more EMPs, it can level the playing field somewhat better by at least providing a glimpse of what's happening around the mech during an ion/EMP onslaught

Time-Green replied:
Okay

Time-Green commented on code/modules/vehicles/mecha/mecha_movement.dm:
```diff
@@ -148,3 +148,27 @@
+	. = ..()
+	//Only bother updating the camera if we actually managed to move
+	if(. && chassis_camera)
+		update_camera_location(destination)
```
Why do you check for forceMove? Moved is already called after a succesful forceMove

Time-Green commented on code/modules/vehicles/mecha/mecha_movement.dm:
```diff
@@ -148,3 +148,27 @@
+		update_camera_location(destination)
+
+/obj/vehicle/sealed/mecha/proc/do_camera_update(oldLoc)
+	if(!QDELETED(chassis_camera) && oldLoc != get_turf(src))
```
Why all the qdeleted checks? If this was an issue you ran into, you should make sure to clear the camera on mech deletion/destruction

Time-Green commented on code/modules/vehicles/mecha/mecha_movement.dm:
```diff
@@ -148,3 +148,27 @@
+	if(. && chassis_camera)
+		update_camera_location(destination)
+
+/obj/vehicle/sealed/mecha/proc/do_camera_update(oldLoc)
```
Autodoc this

Time-Green commented on code/modules/vehicles/mecha/mecha_movement.dm:
```diff
@@ -148,3 +148,27 @@
+	updating = FALSE
+
+#define MECHA_CAMERA_BUFFER 0.5 SECONDS
+/obj/vehicle/sealed/mecha/proc/update_camera_location(oldLoc)
```
Autodoc this

Time-Green commented on code/modules/vehicles/mecha/mecha_movement.dm:
```diff
@@ -148,3 +148,27 @@
+	oldLoc = get_turf(oldLoc)
+	if(!QDELETED(chassis_camera) && !updating && oldLoc != get_turf(src))
+		updating = TRUE
+		addtimer(CALLBACK(src, PROC_REF(do_camera_update), oldLoc), MECHA_CAMERA_BUFFER)
```
Why do we wait half a second for this?

Sealed101 replied:
most of the above code is just copypasted from how borgs do it i have no excuse

Time-Green replied:
Eh, ugh... I'm afraid there's probably some cursed reason for it then. It's fiiiiine

Sealed101 committed:
some review tidbits

Sealed101 committed:
moving cameras pass unique update timer
silicon and mech cameras pass 0.5 sec buffer timer down to hasChanged
wipe away forceMove checks from silicon/mecha
fix up some more stuff in camera console
bring up secureye (laptop/modconsole app) to use new changes
hi :)

Sealed101 commented:
as per Kylerace's suggestions, silicon and mecha cameras will now pass their own update buffer to `hasChanged` and `update`'s buffer timer. i've set both cameras to pass a 0.5 second buffer, it seems to be working pretty well. mechs with an EMPd camera or a very forcefully thrown borg in theory can escape the camera's feed however.

https://user-images.githubusercontent.com/75863639/204329262-3256a63b-6560-48e4-b54d-ed7ada441695.mp4

camera consoles as well as the SecurEye app don't update their feeds as the cameras do, they depend on tgui's process, i don't think i should touch that honestly

posting this here before it vanishes in discord's bowels

![image](https://user-images.githubusercontent.com/75863639/204330660-910c5bf3-f5c1-48b8-9138-22e35c24e39f.png)

Sealed101 committed:
forgor to comment these two

Sealed101 committed:
this should do it

Sealed101 commented:
yea i think it's ready

Time-Green commented on code/game/machinery/camera/presets.dm:
```diff
@@ -66,6 +66,38 @@
 
 	c_tag = "[format_text(camera_area.name)] #[number]"
 
+/obj/machinery/camera/exosuit
```
Can you autodoc tje class?

Time-Green committed:
Autodocs /obj/machinery/camera/exosuit class
